cpp/src/arrow/c/abi.h

```diff
@@ -60,6 +63,11 @@ struct ArrowArray {
   void* private_data;
 };
 
+#endif  // ARROW_C_DATA_H
```

Comment:
It seems that this mixes `extern "C" {...}` and `#ifndef ... #endif`:

```c
#ifndef ARROW_C_DATA_H
#define ARROW_C_DATA_H

#ifdef __cplusplus
extern "C" {
#endif

#endif  // ARROW_C_DATA_H

#ifndef ARROW_C_STREAM_H
#define ARROW_C_STREAM_H

#ifdef __cplusplus
}
#endif

#endif  // ARROW_C_STREAM_H
```

I think that it may not be safe. How about avoiding the mix?

```c
#ifndef ARROW_C_DATA_H
#define ARROW_C_DATA_H

#ifdef __cplusplus
extern "C" {
#endif

#ifdef __cplusplus
}
#endif

#endif  // ARROW_C_DATA_H

#ifndef ARROW_C_STREAM_H
#define ARROW_C_STREAM_H

#ifdef __cplusplus
extern "C" {
#endif

#ifdef __cplusplus
}
#endif

#endif  // ARROW_C_STREAM_H
```

Reply:
Oops, indeed, you're right. Thanks for catching this!